R/tm_g_heat_bygrade.R

```diff
@@ -352,62 +350,75 @@ srv_g_heatmap_bygrade <- function(id,
     })
 
     output_q <- reactive({
-      iv_len <- shinyvalidate::InputValidator$new()
-      anno_var <- input$anno_var
-      iv_len$add_rule("anno_var", function(x) if (length(x) > 2) "Please include no more than 2 annotation variables.")
-      iv_len$enable()
-      validate(need(iv_len$is_valid(), "Misspecification error: please observe red flags in the encodings."))
-      validate(need(iv$is_valid(), "Misspecification error: please observe red flags in the encodings."))
-
       ADSL <- data[[sl_dataname]]() # nolint
       ADEX <- data[[ex_dataname]]() # nolint
       ADAE <- data[[ae_dataname]]() # nolint
+      if (isTRUE(input$plot_cm)) {
+        ADCM <- data[[cm_dataname]]() # nolint
+      }
 
-      validate(need(nrow(ADSL) > 0, "Please select at least one subject"))
-      validate(need(
-        input$ongo_var %in% names(ADEX),
-        paste("Study Ongoing Status must be a variable in", ex_dataname, sep = " ")
-      ))
+      teal::validate_has_data(ADSL, min_nrow = 0, msg = sprintf("%s contains no data", sl_dataname))
 
-      validate(need(
-        checkmate::test_logical(ADEX[[input$ongo_var]], min.len = 1, any.missing = FALSE),
+      iv <- shinyvalidate::InputValidator$new()
+      iv$add_rule("id_var", shinyvalidate::sv_required(
+        message = "ID Variable is required"
+      ))
+      iv$add_rule("visit_var", shinyvalidate::sv_required(
+        message = "Visit Variable is required"
+      ))
+      iv$add_rule("ongo_var", shinyvalidate::sv_required(
+        message = "Study Ongoing Status Variable is required"
+      ))
+      iv$add_rule("ongo_var", shinyvalidate::sv_in_set(
+        set = names(ADEX),
+        message_fmt = sprintf("Study Ongoing Status must be a variable in %s", ex_dataname)
+      ))
+      iv$add_rule("ongo_var", ~ if (!is.logical(ADEX[[req(.)]])) {
```

Comment:
Having the req() there is quite confusing? As you've put sv_required above this is it needed?

Reply:
`req` removed